File: c++/medium/238.product_array_except_self/product_of_arra_except_self.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        vector<int> answer(n);
        vector<int> leftPrefixButSelf(n);
        vector<int> rightPrefix(n);

        leftPrefixButSelf[0] = 1;
        rightPrefix[0] = nums[n-1];

        int previous = nums[0];
        for(int i = 1; i < n; i++) {
            rightPrefix[i] = rightPrefix[i-1] * nums[n-1-i];
            leftPrefixButSelf[i] = previous;
            previous *= nums[i];
        }

        for(int i = 0; i < n-1; i++) {
            answer[i] = leftPrefixButSelf[i] * rightPrefix[n - (i+2)];
        }

        answer[n-1] = leftPrefixButSelf[n-1];
        return answer;
    }
};

File: python/medium/238.product_array_except_self/product_of_arra_except_self.py
from typing import List

class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        answer = [0] * n

        left_prefix = [0] * n
        right_prefix = [0] * n

        left_prefix[0] = 1
        right_prefix[0] = nums[n-1]

        multiplication = nums[0]

        for i in range(1, n):
            right_prefix[i] = right_prefix[i-1] * nums[n-1-i]
            left_prefix[i] = multiplication
            multiplication *= nums[i]
        
        for i in range(0, n-1):
            answer[i] = left_prefix[i] * right_prefix[n - (i+2)]

        answer[n-1] = left_prefix[n-1]
        return answer
